### PR TITLE
Fix Raspberry Pi model detection on mainline Linux

### PR DESCRIPTION
**Pull request by neuschaefer**
(please do not squash)

### CODE REVIEW

File: lib/gpio.cc [hunk 1]
```diff
@@ -244,25 +244,54 @@ static int ReadFileToBuffer(char *buffer, size_t size, const char *filename) {
...
+  return pi_revision;
+}
+
+static uint32_t read_be32(const uint8_t *p) {
+  return p[0] << 24 | p[1] << 16 | p[2] << 8 | p[3];
+}
+
+static uint32_t ReadRevisionFromDeviceTree() {
+  char buffer[4];
+  if (ReadFileToBuffer(buffer, sizeof(buffer), "/proc/device-tree/system/linux,revision") < 0) {
```
Comment (hzeller) [inline lib/gpio.cc:278]: Note, `ReadFileToBuffer()` is meant to read a text file, so it reads one byte less than in the buffer and appends a nul-termination. So for the 4 byte linux revision to work , you need to give it a 5 byte buffer.

Also, you're checking for negative return value here, i.e. if it _wasn't_ able to read. Wouldn't you want to check for `== 4` here ?

**Comment by hzeller** [comment]
Looks good in general, but I don't think it does what you think it does.

File: lib/gpio.cc (hunk 1 above)
Comment (neuschaefer) [reply lib/gpio.cc:278]: Good points, I'll review the available I/O functions and test again

File: lib/gpio.cc [hunk 2]
```diff
@@ -269,10 +269,26 @@ static uint32_t ReadRevisionFromProcCpuinfo() {
...
 
+static uint32_t read_be32(const uint8_t *p) {
+  return p[0] << 24 | p[1] << 16 | p[2] << 8 | p[3];
+}
+
+static uint32_t ReadRevisionFromDeviceTree() {
+  char buffer[4];
+  if (ReadFileToBuffer(buffer, sizeof(buffer), "/proc/device-tree/system/linux,revision") < 0) {
+    fprintf(stderr, "Failed to read revision from /proc/device-tree\n");
+    return read_be32((const uint8_t *)buffer);
```
Comment (neuschaefer) [inline lib/gpio.cc:280]: Another thing here: I've used a C-style cast, but I'm open to suggestions if you think there's a C++ cast that fits better

**Comment by neuschaefer** [comment]
I think the logic is sound now, and I've checked that the revision is read correctly.

**Review by hzeller — APPROVED** [review]

**Pull request merged**